tfhe/examples/dark_market.rs

```diff
@@ -0,0 +1,303 @@
+    println!(
+        "Calculated total volume to be matched in {:?}",
+        time.elapsed()
+    );
+
+    let fill_orders = |orders: &mut [RadixCiphertextBig]| {
+        let mut volume_left_to_transact = total_volume.clone();
+        for mut order in orders.iter_mut() {
+            let mut filled_amount = server_key.smart_min(&mut volume_left_to_transact, &mut order);
+            server_key.smart_sub_assign(&mut volume_left_to_transact, &mut filled_amount);
+            *order = filled_amount;
+        }
+    };
+
+    println!("Filling orders...");
+    let time = Instant::now();
+    fill_orders(sell_orders);
+    fill_orders(buy_orders);
+    println!("Filled orders in {:?}ms", time.elapsed());
+}
+
+/// FHE implementation of the volume matching algorithm.
+///
+/// This version of the algorithm utilizes parallelization to speed up the computation.
+///
+/// Matches the given encrypted [sell_orders] with encrypted [buy_orders] using the given [server_key].
+/// The amount of the orders that are successfully filled is written over the original order count.
+fn volume_match_fhe_parallelized(
+    sell_orders: &mut [RadixCiphertextBig],
+    buy_orders: &mut [RadixCiphertextBig],
+    server_key: &ServerKey,
+) {
+    // Calculate the element sum of the given vector in parallel
+    let parallel_vector_sum = |vec: &mut [RadixCiphertextBig]| {
+        vec.to_vec().into_par_iter().reduce(
+            || server_key.create_trivial_zero_radix(NUMBER_OF_BLOCKS),
+            |mut acc: RadixCiphertextBig, mut ele: RadixCiphertextBig| {
+                server_key.smart_add_parallelized(&mut acc, &mut ele)
+            },
+        )
+    };
+
+    println!("Calculating total sell and buy volumes...");
+    let time = Instant::now();
+    // Total sell and buy volumes can be calculated in parallel because they have no dependency on each other.
+    let (mut total_sell_volume, mut total_buy_volume) = rayon::join(
+        || parallel_vector_sum(sell_orders),
+        || parallel_vector_sum(buy_orders),
+    );
+    println!(
+        "Total sell and buy volumes are calculated in {:?}",
+        time.elapsed()
+    );
+
+    println!("Calculating total volume to be matched...");
+    let time = Instant::now();
+    let total_volume =
+        server_key.smart_min_parallelized(&mut total_sell_volume, &mut total_buy_volume);
+    println!(
+        "Calculated total volume to be matched in {:?}",
+        time.elapsed()
+    );
+
+    let fill_orders = |orders: &mut [RadixCiphertextBig]| {
+        let mut volume_left_to_transact = total_volume.clone();
+        for mut order in orders.iter_mut() {
+            let mut filled_amount =
+                server_key.smart_min_parallelized(&mut volume_left_to_transact, &mut order);
+            server_key
+                .smart_sub_assign_parallelized(&mut volume_left_to_transact, &mut filled_amount);
+            *order = filled_amount;
+        }
+    };
```

Comment:
fill_orders here is sequential, we know there is at most one partially filled order, I feel like there might be a way to decide whether a chunk of orders is fullfilled by comparing partial sums of the X first orders with the total volume to transact

the partial sums should be computable in parallel (more work but with a machine with a lot of threads doing more work to improve latency should be fine), to avoid having the same sequential characteristic as the current implementation, there may be a way to write this in a divide and conquer fashion deciding which state is the correct state

For example:

Sell orders:
[2, 3, 4, 5, 6]

Buy orders:
[1, 1, 2, 1, 1]

total volumes : 6

partial Buy orders sum (computed with tree reduction and in parallel):
[1, 2, 4, 5, 6]

then the output filled amount with equivalent fhe ops:
(the 6 in the <= comparison is the total volume)
`[(1 <= 6) * 1 + /* manage partial fill or 0 fill */, ..., (6<= 6) * 6 + /* manage partial fill or 0 fill */]`

this way the final fill might be done in parallel and the most costly part should be the partial sums computations which (for infinite thread) should only take as long as it takes the compute the total sum in a tree like fashion

Reply:
Great suggestion! I updated in the way you recommended.

Reply:
Good news, runtime went down to 124 seconds with the parallel prefix sum trick on our m6i.metal benchmark instance !

I'll test enabling avx512

Reply:
and with avx512 108 s not too bad !

Reply:
That's great news!